whetstone.py

```diff
@@ -32,22 +59,482 @@ def _authorize(self):
+        """Convert date columns to the actual date time."""
+        date_types = {col: "datetime64[ns]" for col in df.columns if col in self.dates}
+        if date_types:
+            df = df.astype(date_types)
+        return df
+
+
+class Users(Whetstone):
+    def __init__(self, sql, qa=False):
+        super().__init__(sql)
+        self.columns = [
+            "_id",
+            "internalId",
+            "activeDistrict",
+            "archivedAt",
+            "created",
+            "email",
+            "inactive",
+            "lastActivity",
+            "lastModified",
+            "locked",
+            "name",
+            "first",
+            "last",
+            "school",
+            "course",
+            "coach",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {"Users": []}
+        for record in records:
+            record["school"] = record.get("defaultInformation").get("school")
+            record["course"] = record.get("defaultInformation").get("course")
+            user = {k: v for (k, v) in record.items() if k in self.columns}
+            models["Users"].append(user)
+        return models
+
+
+class Schools(Whetstone):
+    def __init__(self, sql, qa=False):
+        super().__init__(sql)
+        self.columns = [
+            "_id",
+            "internalId",
+            "name",
+            "abbreviation",
+            "archivedAt",
+            "principal",
+            "gradeSpan",
+            "lowGrade",
+            "highGrade",
+            "district",
+            "phone",
+            "address",
+            "city",
+            "cluster",
+            "region",
+            "state",
+            "zip",
+            "lastModified",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {"Schools": [], "ObservationGroups": [], "ObservationGroupMembers": []}
+        for record in records:
+            record_id = record.get("_id")
+            school = {k: v for (k, v) in record.items() if k in self.columns}
+            models["Schools"].append(school)
+            groups = record.get("observationGroups")
+            if groups:
+                for group in groups:
+                    observers = self._extract_group_members(
+                        group, "observers", record_id
+                    )
+                    models["ObservationGroupMembers"].extend(observers)
+                    observees = self._extract_group_members(
+                        group, "observees", record_id
+                    )
+                    models["ObservationGroupMembers"].extend(observees)
+
+                groups = [
+                    {
+                        "id": item.get("_id"),
+                        "name": item.get("name"),
+                        "school": record_id,
+                        "lastModified": item.get("lastModified"),
+                    }
+                    for item in groups
+                ]
```

Comment:
Multi-line list comprehensions are pretty uncommon, I think. This probably reads more easily as a for loop adding dictionaries to a list? Performance should be effectively the same.

Reply:
@zkagin I'd prefer to avoid the loop here. I find this much more concise, but I see your point. What is I pull this logic out into it's own function and then apply it more like a map?